File: 0207-course-schedule/0207-course-schedule.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        courses = collections.defaultdict(list)
        
        def is_cycle(node, visited, tracker):
            visited[node] = True
            tracker[node] = True
            
            for i in courses[node]:
                if i not in visited and is_cycle(i, visited, tracker):
                    return True
                elif i in tracker:
                    return True
            tracker.pop(node)
            return False
        
        
        for after, before in prerequisites:
            courses[before].append(after)
        
        visited = {}
        
        for i in range(numCourses):
            if i not in visited and is_cycle(i, visited, {}):
                return False
        
        return True

File: 0207-course-schedule/0207-course-schedule.java
class Solution {
    List<List<Integer> > graph;
    boolean[] visited;
    
    public boolean canFinish(int numCourses, int[][] prerequisites) {
        graph = new ArrayList<>();
        visited = new boolean[numCourses];
        
        for (int i = 0; i < numCourses; i++) {
            graph.add(new ArrayList<>());
        }
        
        for (int[] courses : prerequisites) {
            graph.get(courses[1]).add(courses[0]);
        }
        
        for (int i = 0; i < numCourses; i++) {
            if(!visited[i] && hasCycle(i, new HashSet<Integer>())) {
                return false;
            }
        }
        
        return true;
    }
    
    public boolean hasCycle(int node, Set<Integer> tracker) {
        tracker.add(node);
        visited[node] = true;
        
        for (Integer next : graph.get(node)) {
            if (!visited[next] && hasCycle(next, tracker)) {
                return true;
            } else if(tracker.contains(next)) {
                return true;
            }            
        }
        tracker.remove(node);
        
        return false;
    }
}